File: Programming/JavaScript/CombinationSum/combinationSum.js
const combinationSum = (candidates, target) => {
  const result = [];
  const permute = (arr = [], sum = 0, idx = 0) => {
    if (sum > target) return;
    if (sum === target) result.push(arr);

    for (let i = idx; i < candidates.length; i++) {
      permute([...arr, candidates[i]], sum + candidates[i], i);
    }
  };
  permute();
  return result;
};

combinationSum([2, 3, 6, 7], 7);        // returns [ [ 2, 2, 3 ], [ 7 ] ]
combinationSum([2, 3, 5], 8);           // returns [ [ 2, 2, 2, 2 ], [ 2, 3, 3 ], [ 3, 5 ] ]
combinationSum([2], 1);                 // returns [ ]
combinationSum([1], 1);                 // returns [ [ 1 ] ]
combinationSum([1], 2);                 // returns [ [ 1, 1 ] ]

File: Programming/C++/CombinationSum/CombinationSum.cpp
class Solution
{
public:
    vector<vector<int>> vec;

    void helper(vector<int> &nums, int key, vector<int> &currentItems, int idx = 0)
    {
        if (key == 0)
        {
            vec.push_back(currentItems);
            return;
        }

        if (key < 0)
            return;
        if (idx < nums.size())
        {
            currentItems.push_back(nums[idx]);
            helper(nums, key - nums[idx], currentItems, idx); // if not same element then idx+1
            currentItems.pop_back();
            helper(nums, key, currentItems, idx + 1);
        }
    }

    vector<vector<int>> combinationSum(vector<int> &candidates, int target)
    {
        vector<int> res;
        helper(candidates, target, res);

        return vec;
    }
};